pkg/api/pluginproxy/access_token_provider_test.go

```diff
@@ -91,4 +98,172 @@ func TestAccessToken(t *testing.T) {
+			Path:   "pathwithtokenauth1",
+			Url:    "",
+			Method: "GET",
+			TokenAuth: &plugins.JwtTokenAuth{
+				Url: server.URL + "/oauth/token",
+				Scopes: []string{
+					"https://www.testapi.com/auth/monitoring.read",
+					"https://www.testapi.com/auth/cloudplatformprojects.readonly",
+				},
+				Params: map[string]string{
+					"grant_type":    "client_credentials",
+					"client_id":     "{{.JsonData.client_id}}",
+					"client_secret": "{{.SecureJsonData.client_secret}}",
+					"audience":      "{{.JsonData.audience}}",
+					"client_name":   "datasource_plugin",
+				},
+			},
+		}
+
+		templateData := templateData{
+			JsonData: map[string]interface{}{
+				"client_id": "my_client_id",
+				"audience":  "www.example.com",
+			},
+			SecureJsonData: map[string]string{
+				"client_secret": "my_secret",
+			},
+		}
+
+		var authCalls int
+		apiHandler.HandleFunc("/oauth/token", func(w http.ResponseWriter, req *http.Request) {
+			err := json.NewEncoder(w).Encode(token)
+			if err != nil {
+				t.Error(err)
+			}
+			authCalls++
+		})
+
+		Convey("Should parse token, with different fields and types", func() {
+			type tokenTestDescription struct {
+				desc              string
+				expiresIn         interface{}
+				expiresOn         interface{}
+				expectedExpiresOn int64
+			}
+
+			mockTimeNow(time.Now().Unix())
+			defer resetTimeNow()
+			provider := newAccessTokenProvider(&models.DataSource{}, pluginRoute)
+
+			testCases := []tokenTestDescription{
+				{
+					desc:              "token with expires_in in string format",
+					expiresIn:         "3600",
+					expiresOn:         nil,
+					expectedExpiresOn: timeNow().Unix() + 3600,
+				},
+				{
+					desc:              "token with expires_in in int format",
+					expiresIn:         5200,
+					expiresOn:         nil,
+					expectedExpiresOn: timeNow().Unix() + 5200,
+				},
+				{
+					desc:              "token with expires_on in string format",
+					expiresOn:         timeNow().Add(86 * time.Minute).Unix(),
+					expiresIn:         nil,
+					expectedExpiresOn: timeNow().Add(86 * time.Minute).Unix(),
+				},
+				{
+					desc:              "token with expires_on in int format",
+					expiresOn:         timeNow().Add(156 * time.Minute).Unix(),
+					expiresIn:         nil,
+					expectedExpiresOn: timeNow().Add(156 * time.Minute).Unix(),
+				},
+				{
+					desc:              "token with both expires_on and expires_in, should prioritize expiresOn",
+					expiresIn:         5200,
+					expiresOn:         timeNow().Add(1 * time.Hour).Unix(),
+					expectedExpiresOn: timeNow().Add(1 * time.Hour).Unix(),
+				},
+			}
+			for _, testCase := range testCases {
+				clearTokenCache()
+				// reset the httphandler counter
+				authCalls = 0
+
+				token = map[string]interface{}{
+					"access_token":  "2YotnFZFEjr1zCsicMWpAA",
+					"token_type":    "example",
+					"expires_in":    testCase.expiresIn,
+					"expires_on":    testCase.expiresOn,
```

Comment:
I would prefer to include `expires_in` and `expires_on` only if `testCase.expiresIn` and `testCase.expiresOn` are not `nil` in order to demonstrate the case that either of two can be missing.